python/pyspark/ml/tests/test_util.py

```diff
@@ -15,63 +15,188 @@
+def test_function(x: float, y: float) -> float:
+    return x**2 + y**2
+
+class TestFunctionPickler(unittest.TestCase):
+
+    def check_if_test_function_pickled(self, f, og_fn, output_value, *arguments, **key_word_args):
+        fn, args, kwargs = cloudpickle.load(f)
+        self.assertEqual(fn, og_fn)
+        self.assertEqual(args, arguments)
+        self.assertEqual(kwargs, key_word_args)
+        fn_output = fn(*args, **kwargs)
+        self.assertEqual(fn_output, output_value)
+
+    def test_pickle_fn_and_save(self):
+        x, y = 1, 3 # args of test_function
+        tmp_dir = "silly_goose"
+        os.mkdir(tmp_dir)
+        file_path_to_save = "silly_bear"
+        with self.subTest(msg="See if it pickles correctly if no file_path or save_dir are specified"):
+            pickled_fn_path = FunctionPickler.pickle_fn_and_save(test_function, "", "", x, y)
+            with open(pickled_fn_path, "rb") as f:
+                self.check_if_test_function_pickled(f, test_function, 10, x, y)
+            os.remove(pickled_fn_path)
+        with self.subTest(msg="See if pickles correctly and uses file path given as argument"):
+            pickled_fn_path = FunctionPickler.pickle_fn_and_save(test_function, file_path_to_save, "", x, y)
+            self.assertEqual(pickled_fn_path, file_path_to_save)
+            with open(pickled_fn_path, "rb") as f:
+                self.check_if_test_function_pickled(f, test_function, 10, x, y)
+            os.remove(pickled_fn_path)
+        with self.subTest(msg="See if pickles correctly and uses file path despite save_dir being specified"):
+            pickled_fn_path = FunctionPickler.pickle_fn_and_save(test_function, file_path_to_save, tmp_dir, x, y)
+            self.assertEqual(pickled_fn_path, file_path_to_save)
+            with open(pickled_fn_path, "rb") as f:
+                self.check_if_test_function_pickled(f, test_function, 10, x, y)
+            os.remove(pickled_fn_path)
+
+        os.rmdir(tmp_dir)
+
+    def test_getting_output_from_pickle_file(self):
+        a, b = 2, 0
+        pickle_fn_file = FunctionPickler.pickle_fn_and_save(test_function, "", "", a, b)
+        fn, args, kwargs = FunctionPickler.get_func_output(pickle_fn_file)
+        self.assertEqual(fn, test_function)
+        self.assertEqual(len(args), 2)
+        self.assertEqual(len(kwargs), 0)
+        self.assertEqual(args[0], a)
+        self.assertEqual(args[1], b)
+        self.assertEqual(fn(*args, **kwargs), 4)
+        os.remove(pickle_fn_file)
+    
+    @contextmanager
+    def create_reference_file(self, body: str, prefix: str = "", suffix: str = "", fname: str = "reference.py") -> Iterator[None]:
+        try:
+            with open(fname, "w") as f:
+                if prefix != "":
+                    f.write(prefix)
+                f.write(body)
+                if suffix != "":
+                    f.write(suffix)
+            yield
+        finally:
+            os.remove(fname)
```

Comment:
Why immediately remove it?

Reply:
To clean up the reference file. Because it's a context manager, it'll only clean up upon exiting the context, so the file would've served its purpose for the tests.

Reply:
Oh ok thanks didn't notice the @contextmanager